File: dijkstras/dijkstras.go
package main

import (
	"fmt"
	"math"
)

var graph map[string]map[string]float64
var costs map[string]float64
var parents map[string]string
var processed []string

func main() {
	// the graph
	graph = make(map[string]map[string]float64)
	graph["start"] = make(map[string]float64)
	graph["start"]["a"] = 6
	graph["start"]["b"] = 2

	graph["a"] = make(map[string]float64)
	graph["a"]["fin"] = 1

	graph["b"] = make(map[string]float64)
	graph["b"]["a"] = 3
	graph["b"]["fin"] = 5

	graph["fin"] = make(map[string]float64)

	// the costs table
	costs = make(map[string]float64)
	costs["a"] = 6
	costs["b"] = 2
	costs["fin"] = math.Inf(1)

	// the parents table
	parents = make(map[string]string)
	parents["a"] = "start"
	parents["b"] = "start"
	parents["fin"] = ""
	node := find_lowest_cost_node(costs)

	for node != ""{
		cost := costs[node]
		neighbors := graph[node]
		for n := range neighbors{
			new_cost := cost + neighbors[n]
			if new_cost < costs[n]{
				costs[n] = new_cost
				parents[n] = n
			}
		} 
		processed = append(processed, node)
		node = find_lowest_cost_node(costs)
	}
	fmt.Println(costs)



}

func find_lowest_cost_node(costs map[string]float64) (node string){
	lowest_cost := math.Inf(1)
	lowest_cost_node := ""
	for n := range costs{
		if costs[n] < lowest_cost && !contains(processed, n){
			lowest_cost = costs[n]
			lowest_cost_node = n
		}
	} 
	return lowest_cost_node
}

func contains(li []string, n string) bool{
	for _, v := range li{
		if v == n{
			return true
		}
	}
	return false
}

File: dijkstras/dijkstras.py
# graph = dict()

# graph["start"] = {"a":6, "b":2}

# graph["a"] = {"fin":1}

# graph["b"] = {"fin":5, "a":3}
# // ssss
# graph["fin"] = {}

inf = float("inf")

graph = {"start":{"a":6, "b":2}, "a":{"fin":1}, "b":{"fin":5, "a":3}, "fin":{}}

costs = {"a": 6, "b":2, "fin": inf}

parent = {"a":"start", "b":"start", "fin":None}

processed = []

def find_lowest_cost_node(costs):
    lowest_cost = float("inf")
    lowest_cost_node = None
    
    for node in costs:
        cost = costs[node]
        if cost < lowest_cost and node not in processed:
            lowest_cost = cost
            lowest_cost_node = node
    return lowest_cost_node
def find():
    node = find_lowest_cost_node(costs)

    while node is not None:
        cost = costs[node]
        neighbors = graph[node]
        
        for n in neighbors.keys():
            new_cost = cost + neighbors[n]
            if costs[n]> new_cost:
                costs[n] = new_cost
                parent[n] = node
        processed.append(node)
        node = find_lowest_cost_node(costs)
    print("Cost from the start to each node:")
    print(costs)
    
if __name__ == "__main__":
    find()